pkg/object/ingresscontroller/ingresscontroller.go

```diff
@@ -0,0 +1,239 @@
+		Namespaces   []string         `yaml:"namespaces" jsonschema:"omitempty"`
+		IngressClass string           `yaml:"ingressClass" jsonschema:"omitempty"`
+	}
+)
+
+func init() {
+	supervisor.Register(&IngressController{})
+}
+
+// Category returns the category of IngressController.
+func (ic *IngressController) Category() supervisor.ObjectCategory {
+	return Category
+}
+
+// Kind return the kind of IngressController.
+func (ic *IngressController) Kind() string {
+	return Kind
+}
+
+// DefaultSpec returns the default spec of IngressController.
+func (ic *IngressController) DefaultSpec() interface{} {
+	return &IngressControllerSpec{
+		HTTPServer: &httpserver.Spec{
+			KeepAlive:        true,
+			KeepAliveTimeout: "60s",
+			MaxConnections:   10240,
+		},
+		IngressClass: defaultIngressClass,
+	}
+}
+
+// Init initializes IngressController.
+func (ic *IngressController) Init(superSpec *supervisor.Spec) {
+	ic.superSpec = superSpec
+	ic.spec = superSpec.ObjectSpec().(*IngressControllerSpec)
+	ic.super = superSpec.Super()
+	ic.reload()
+}
+
+// Inherit inherits previous generation of IngressController.
+func (ic *IngressController) Inherit(superSpec *supervisor.Spec, previousGeneration supervisor.Object) {
+	previousGeneration.Close()
+	ic.Init(superSpec)
+}
+
+func (ic *IngressController) reload() {
+	if entity, exists := ic.super.GetSystemController(trafficcontroller.Kind); !exists {
+		panic(fmt.Errorf("BUG: traffic controller not found"))
+	} else if tc, ok := entity.Instance().(*trafficcontroller.TrafficController); !ok {
+		panic(fmt.Errorf("BUG: want *TrafficController, got %T", entity.Instance()))
+	} else {
+		ic.tc = tc
+	}
+
+	ic.namespace = fmt.Sprintf("%s/%s", ic.superSpec.Name(), "ingresscontroller")
+	ic.stopCh = make(chan struct{})
+
+	ic.wg.Add(1)
+	go ic.run()
+}
+
+func (ic *IngressController) run() {
+	defer ic.wg.Done()
+
+	// connect to kubernetes
+	for {
+		k8sClient, err := newK8sClient(ic.spec.MasterURL, ic.spec.KubeConfig)
+		if err == nil {
+			ic.k8sClient = k8sClient
+			break
+		}
+		logger.Errorf("failed to create kubernetes client: %v", err)
+
+		select {
+		case <-ic.stopCh:
+			return
+		case <-time.After(10 * time.Second):
+		}
+	}
+	logger.Infof("successfully connect to kubernetes")
+
+	// watch ingress related resources
+	var stopCh chan struct{}
+	for {
+		// use a new stopCh here to clear goroutines created by watch.
+		// if we use ic.stopCh, they cannot be stopped before the return
+		// of this function, which result in resource leaks
+		stopCh = make(chan struct{})
+		err := ic.k8sClient.watch(ic.spec.Namespaces, stopCh)
+		if err == nil {
+			break
+		}
+		close(stopCh)
+		logger.Errorf("failed to watch ingress related resources: %v", err)
+
+		select {
+		case <-ic.stopCh:
+			return
+		case <-time.After(10 * time.Second):
+		}
+	}
+	logger.Infof("successfully watched ingress related resources")
+
+	// process resource update events
+	var err error
+	for {
+		select {
+		case <-ic.stopCh:
```

Comment:
Should be the sender's responsibility to close the channel?

## Reference
1. https://tour.golang.org/concurrency/4

Reply:
`ic->stopCh` and `stopCh` are two channels.

Reply:
If so, propose to move the stop channel into the `k8sclient` structure as its `watch` API's return value.  The `IngressController` only needs to receive this channel's closing message.